File: my-folder/0073-set-matrix-zeroes/solution.py
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        row_flag = False
        col_flag = False
        for row in range(len(matrix)):
            for col in range(len(matrix[0])):
                if matrix[row][col] == 0:
                    if row == 0:
                        row_flag = True
                    if col == 0:
                        col_flag = True
                    elif row != 0 and col != 0:
                        matrix[row][0] = 0
                        matrix[0][col] = 0
        
        for row in range(1,len(matrix)):
            for col in range(1, len(matrix[0])):
                if matrix[row][0] == 0 or matrix[0][col] == 0:
                    matrix[row][col] = 0

        if row_flag:
            matrix[0] = [0] * len(matrix[0])
        
        if col_flag:
            for row in range(len(matrix)):
                matrix[row][0] = 0
            



        
        

File: my-folder/0073-set-matrix-zeroes/solution.java
class Solution {
    public void setZeroes(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;
        
        // we track the intersection of row0 and col0 seperately using this flag 
        int col0 = 1;

       // update all the first row and col positions to 0 when we encounter a zero 
        for(int row = 0; row < rows; row++){
           
            if(matrix[row][0] == 0) col0 = 0;

            for(int col = 1; col < cols; col++){
                if(matrix[row][col] == 0){
                    matrix[row][0] = 0;
                    matrix[0][col] = 0;
                }
            }
        }

        // update from last index position to avoid overwriting markers  
        for(int row = rows - 1; row >= 0; row--){
            for(int col = cols - 1; col >= 1; col--){
                if(matrix[row][0] == 0 || matrix[0][col] == 0){
                    matrix[row][col] = 0;
                }
            }

            if(col0 == 0) matrix[row][0] = 0;
        }
    }
}


